sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -1066,12 +1075,144 @@ async Task performProcessing()
+            // of processing for claimed partitions.
+            //
+            // Ensure that any errors observed, whether fatal or non-fatal, are surfaced to the exception handler in
+            // a fire-and-forget manner.  The processor does not assume responsibility for observing or surfacing exceptions
+            // that may occur in the handler, so the associated task is intentionally unobserved.
+
+            try
+            {
+                var connection = CreateConnection();
+                await using var connectionAwaiter = connection.ConfigureAwait(false);
+
+                var retryPolicy = Options.RetryOptions.ToRetryPolicy();
+                var cycleDuration = new Stopwatch();
+                var partitionIds = default(string[]);
+
+                while (!cancellationToken.IsCancellationRequested)
+                {
+                    cycleDuration.Restart();
+
+                    try
+                    {
+                        partitionIds = await connection.GetPartitionIdsAsync(retryPolicy, cancellationToken).ConfigureAwait(false);
+                    }
+                    catch (Exception ex) when (ex.IsNotType<TaskCanceledException>())
+                    {
+                        // Logging for exceptions with the service operation are responsibility of the connection.
+
+                        _ = InvokeOnProcessingErrorAsync(ex, null, Resources.OperationGetPartitionIds, CancellationToken.None);
+                        partitionIds = default;
+                    }
+
+                    // Perform a load balancing cycle only in the case that partitions were successfully read.
+
+                    if (partitionIds != default)
+                    {
+                        var claimedOwnership = default(EventProcessorPartitionOwnership);
+
+                        try
+                        {
+                            claimedOwnership = await LoadBalancer.RunLoadBalancingAsync(partitionIds, cancellationToken).ConfigureAwait(false);
+                        }
+                        catch (EventHubsException ex)
+                            when (Resources.OperationClaimOwnership.Equals(ex.GetFailureOperation(), StringComparison.InvariantCultureIgnoreCase))
+                        {
+                            // If a specific partition was associated with the failure, it is carried by a well-known data member of the exception.
+
+                            var partitionId = ex.GetFailureData<string>();
+
+                            var partition = (partitionId ?? string.Empty) switch
+                            {
+                                string id when (id == string.Empty) => null,
+                                string _ when (ActivePartitionProcessors.TryGetValue(partitionId, out var partitionProcessor)) => partitionProcessor.Partition,
+                                _ => new TPartition { PartitionId = partitionId }
```

Comment:
The first and second are far more likely; the third is possible but unlikely.   Here's the breakdown, as I see it:

- If the load balancer encounters an error that isn't specific to a partition, then we would hit the first case.  This could be in querying the storage provider for a list of ownerships or similar.

- If the load balancer encounters an error renewing ownership for a partition that is owned and which we're currently processing, we'd hit the second case.  This could be in requesting the storage provider update the ownership records to extend a partition that we already own.

- If the load balancer is making a decision about a partition that we would like to claim and believes that it is owned, but processing hasn't fully started yet, or the attempt to start processing the partition failed, then we would hit the third case.

Reply:
> If the load balancer is making a decision about a partition that we would like to claim and believes that it is owned, but processing hasn't fully started yet, or the attempt to start processing the partition failed, then we would hit the third case.

In this case, do you think there is value in passing as "partially initialized" `TPartition` in, vs say `null`?  I get that we're hiding a small amount of information here but I wonder if that leads to a nicer overall model.

Just something to noodle over.

Reply:
It's a fair question, and I don't know that I have a solid answer.  Passing the partially initialized object does allow the partition identifier to be communicated, which may offer insight into partition-specific issues or help diagnose that there is an issue with the load balancer putting too much contention on ownership for a single partition.

On the other hand, you're not wrong in that it is potentially confusing and may not provide the full set of context for a custom type.   I'm kind of banking on somewhat of an understanding given that the initialize handler won't have fired for the partition and that is where developers would perform the initialization of the custom type. 

I'm certainly open to further discussion and better ideas.   In the meantime, I'll add some additional comments to the error handler to try and clarify that the partition may only be minimally initialized when called.

Reply:
> I'll add some additional comments to the error handler to try and clarify that the partition may only be minimally initialized when called.

Thanks!